src/Workspaces/Core/Portable/Workspace/Solution/CachingAnalyzerConfigSet.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class CachingAnalyzerConfigSet
+    {
+        private readonly ConcurrentDictionary<string, AnalyzerConfigOptionsResult> _cache = new ConcurrentDictionary<string, AnalyzerConfigOptionsResult>();
+        private readonly Func<string, AnalyzerConfigOptionsResult> _computeFunction;
+        private readonly AnalyzerConfigSet _underlyingSet;
+
+        public CachingAnalyzerConfigSet(AnalyzerConfigSet underlyingSet)
+        {
+            _underlyingSet = underlyingSet;
+            _computeFunction = _underlyingSet.GetOptionsForSourcePath;
+        }
+
+        public AnalyzerConfigOptionsResult GetOptionsForSourcePath(string sourcePath)
+        {
+            return _cache.GetOrAdd(sourcePath, _computeFunction);
```

Comment:
hrmm.. why not have the cachign in teh ANalyzerConfigSet itself?

Reply:
The compiler also calls into AnalyerConfigSet and does it's own (different) caching atop it, so I didn't want to push that right away, at least not as something we'll be taking as a hotfix.

Reply:
And more generally: I'm not sure if we'll want smarter caching policies here to limit memory usage, so I think we're best off keeping caching managed by the caller.